File: functions/screen.py
#!/usr/bin/env python3
"""
Netlify Function: /api/screen
輸入: POST JSON { "symbols": "AAPL NVDA TSLA" }  或 GET ?q=AAPL,NVDA
輸出: JSON { "rows": [...], "bench_ret3m": float, "error": str|null }

server 端拉 Yahoo chart API (避開瀏覽器 CORS)，用純 stdlib 計指標 + 評分。
評分邏輯移植自 Stanley 嘅 us_screener.py (max 8)：
  +2 黃金交叉 (MA50 > MA200)
  +1 RSI(14) 健康 (40 <= RSI <= 70)
  +2 MACD 牛 (MACD line > signal)
  +1 相對強度: 3mo return > 基準(SPY) 3mo return
  +1 水位: -25% <= price vs 52w high < 0
  +1 獎勵: 水位理想 band -15%..-3%
風險位: ATR(14) 止損 (-2xATR)，穿 MA200 = invalidation flag
"""

import json
import urllib.request
from urllib.parse import urlparse, parse_qs

# ───────────────────────── 指標 (純 stdlib) ─────────────────────────
def fetch_yahoo(ticker, range_="2y", interval="1d"):
    url = (f"https://query1.finance.yahoo.com/v8/finance/chart/{ticker}"
           f"?range={range_}&interval={interval}")
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=20) as r:
        data = json.load(r)
    res = data["chart"]["result"][0]
    q = res["indicators"]["quote"][0]
    closes = q["close"]; highs = q["high"]; lows = q["low"]
    valid = [(c, h, l) for c, h, l in zip(closes, highs, lows)
             if c is not None and h is not None and l is not None]
    return [v[0] for v in valid], [v[1] for v in valid], [v[2] for v in valid]


def sma(arr, n):
    if len(arr) < n:
        return None
    return sum(arr[-n:]) / n


def rsi(arr, period=14):
    if len(arr) < period + 1:
        return None
    gains = []; losses = []
    for i in range(1, len(arr)):
        d = arr[i] - arr[i-1]
        gains.append(max(d, 0)); losses.append(max(-d, 0))
    ag = sum(gains[-period:]) / period
    al = sum(losses[-period:]) / period
    if al == 0:
        return 100.0
    return 100 - 100 / (1 + ag / al)


def macd(arr):
    def ema(vals, span):
        k = 2 / (span + 1); out = [vals[0]]
        for v in vals[1:]:
            out.append(v * k + out[-1] * (1 - k))
        return out
    e12 = ema(arr, 12); e26 = ema(arr, 26)
    line = [a - b for a, b in zip(e12, e26)]
    sig = ema(line, 9)
    return line[-1], sig[-1]


def atr(highs, lows, closes, period=14):
    trs = []
    for i in range(1, len(closes)):
        tr = max(highs[i] - lows[i],
                 abs(highs[i] - closes[i-1]),
                 abs(lows[i] - closes[i-1]))
        trs.append(tr)
    if len(trs) < period:
        return None
    return sum(trs[-period:]) / period


# ───────────────────────── 評分 ─────────────────────────
def screen_one(t, bench_ret3m, cfg=None):
    c, h, l = fetch_yahoo(t)
    if not c or len(c) < 210:
        raise ValueError("no data / <min bars")
    ma50 = sma(c, 50); ma200 = sma(c, 200)
    r = rsi(c, 14)
    mline, msig = macd(c)
    cur = c[-1]
    ret3m = (c[-1] / c[-63] - 1) * 100 if len(c) >= 63 else 0.0
    high52 = max(c[-252:]) if len(c) >= 252 else max(c)
    dist_high = (cur / high52 - 1) * 100
    a = atr(h, l, c)
    stop = round(cur - 2.0 * a, 2) if a else None
    below_ma200 = bool(cur < ma200) if ma200 else False

    score = 0
    if ma50 and ma200 and ma50 > ma200: score += 2
    if r and 40 <= r <= 70: score += 1
    if mline > msig: score += 2
    if ret3m > bench_ret3m: score += 1
    if -25.0 <= dist_high < 0: score += 1
    if -15.0 <= dist_high <= -3.0: score += 1

    return {
        "ticker": t.upper(),
        "price": round(cur, 2),
        "golden": bool(ma50 and ma200 and ma50 > ma200),
        "ma50": round(ma50, 2) if ma50 else None,
        "ma200": round(ma200, 2) if ma200 else None,
        "rsi": round(r, 1) if r else None,
        "macd_bull": bool(mline > msig),
        "ret3m": round(ret3m, 1),
        "rel_str": round(ret3m - bench_ret3m, 1),
        "dist_high": round(dist_high, 1),
        "atr": round(a, 2) if a else None,
        "stop_loss": stop,
        "below_ma200": below_ma200,
        "score": score,
        "series": [round(x, 2) for x in c[-120:]],   # 畀 frontend 畫圖 (近120日)
    }


def handler(event, context):
    """Netlify Function entrypoint."""
    try:
        # 解析輸入
        if event.get("httpMethod") == "POST":
            body = json.loads(event.get("body") or "{}")
            raw = body.get("symbols") or body.get("q") or ""
        else:
            qs = parse_qs(urlparse(event.get("rawUrl", event.get("path", ""))).query)
            raw = (qs.get("q") or qs.get("symbols") or [""])[0]
        symbols = [s.strip().upper() for s in raw.replace(",", " ").split() if s.strip()]
        symbols = symbols[:30]  # 防濫用
        if not symbols:
            return _resp(400, {"error": "請提供至少一個 symbol，例如 AAPL NVDA"})

        # 基準 (SPY) 3mo return
        try:
            bc, _, _ = fetch_yahoo("SPY")
            bench_ret3m = (bc[-1] / bc[-63] - 1) * 100 if len(bc) >= 63 else 0.0
        except Exception:
            bench_ret3m = 0.0

        rows = []
        errors = []
        for s in symbols:
            try:
                rows.append(screen_one(s, bench_ret3m))
            except Exception as e:
                errors.append({"ticker": s, "error": str(e)})

        rows.sort(key=lambda r: r["score"], reverse=True)
        return _resp(200, {
            "rows": rows,
            "errors": errors,
            "bench_ret3m": round(bench_ret3m, 1),
            "count": len(rows),
        })
    except Exception as e:
        return _resp(500, {"error": f"server error: {e}"})


def _resp(code, obj):
    return {
        "statusCode": code,
        "headers": {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
        },
        "body": json.dumps(obj, ensure_ascii=False),
    }


# ───────────────────────── Netlify 官方格式 (新版) ─────────────────────────
# 同時支援舊式 AWS Lambda handler(event, context) 做 back-compat。
try:
    from netlify.function import Function, Request, Response

    class Handler(Function):
        def on_post(self, req: Request) -> Response:
            try:
                body = json.loads(req.body) if isinstance(req.body, (str, bytes)) else {}
                if isinstance(req.body, bytes):
                    body = json.loads(req.body.decode())
            except Exception:
                body = {}
            event = {
                "httpMethod": "POST",
                "body": json.dumps(body),
                "rawUrl": "",
            }
            out = handler(event, None)
            return Response(
                body=out["body"],
                status=out["statusCode"],
                headers=out["headers"],
            )

        def on_get(self, req: Request) -> Response:
            qs = req.query_string.decode() if isinstance(req.query_string, bytes) else (req.query_string or "")
            from urllib.parse import parse_qs
            parsed = parse_qs(qs)
            raw = (parsed.get("q") or parsed.get("symbols") or [""])[0]
            event = {"httpMethod": "GET", "body": "", "rawUrl": "/api/screen?" + qs}
            out = handler(event, None)
            return Response(body=out["body"], status=out["statusCode"], headers=out["headers"])
except ImportError:
    # netlify.function 唔喺本地 import 到（正常），keep handler 模式就得
    pass


# ── 本地測試用 (python functions/screen.py) ──
if __name__ == "__main__":
    import sys
    syms = sys.argv[1] if len(sys.argv) > 1 else "AAPL NVDA TSLA"
    class _Ev(dict):
        def __init__(self, **kw):
            super().__init__(**kw)
            self["httpMethod"] = "POST"
            self["body"] = json.dumps({"symbols": syms})
            self["rawUrl"] = ""
    class _Ctx: pass
    out = handler(_Ev(), _Ctx())
    print(out["body"])

File: functions/screen.mjs
#!/usr/bin/env node
/**
 * Netlify Function: /api/screen  (JavaScript/Node — Netlify 只支援 JS/TS function)
 * 輸入: POST JSON { "symbols": "AAPL NVDA TSLA" }  或 GET ?q=AAPL,NVDA
 * 輸出: JSON { "rows": [...], "bench_ret3m": float, "error": str|null }
 *
 * 邏輯移植自原 screen.py：server 端拉 Yahoo chart API（避 CORS），
 * 純 JS 計指標 + 評分（max 8）。
 */

// ───────────────────────── 指標 (純 JS) ─────────────────────────
async function fetchYahoo(ticker, range = "2y", interval = "1d") {
  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${ticker}?range=${range}&interval=${interval}`;
  const res = await fetch(url, { headers: { "User-Agent": "Mozilla/5.0" } });
  if (!res.ok) throw new Error(`yahoo http ${res.status}`);
  const data = await res.json();
  const res0 = data?.chart?.result?.[0];
  if (!res0) throw new Error("no chart result");
  const q = res0.indicators.quote[0];
  const closes = q.close, highs = q.high, lows = q.low, vols = q.volume || [];
  const valid = [];
  for (let i = 0; i < closes.length; i++) {
    if (closes[i] != null && highs[i] != null && lows[i] != null) {
      valid.push([closes[i], highs[i], lows[i], (vols[i] != null ? vols[i] : 0)]);
    }
  }
  return {
    c: valid.map(v => v[0]),
    h: valid.map(v => v[1]),
    l: valid.map(v => v[2]),
    v: valid.map(v => v[3]),
  };
}

function sma(arr, n) {
  if (arr.length < n) return null;
  return arr.slice(-n).reduce((a, b) => a + b, 0) / n;
}

// 滾動 SMA，回同長度 array（前 n-1 位為 null），畀圖表畫 MA 線
function smaSeries(arr, n) {
  const out = new Array(arr.length).fill(null);
  let sum = 0;
  for (let i = 0; i < arr.length; i++) {
    sum += arr[i];
    if (i >= n) sum -= arr[i - n];
    if (i >= n - 1) out[i] = round(sum / n, 2);
  }
  return out;
}

function rsi(arr, period = 14) {
  if (arr.length < period + 1) return null;
  const gains = [], losses = [];
  for (let i = 1; i < arr.length; i++) {
    const d = arr[i] - arr[i - 1];
    gains.push(Math.max(d, 0));
    losses.push(Math.max(-d, 0));
  }
  const ag = gains.slice(-period).reduce((a, b) => a + b, 0) / period;
  const al = losses.slice(-period).reduce((a, b) => a + b, 0) / period;
  if (al === 0) return 100.0;
  return 100 - 100 / (1 + ag / al);
}

function ema(vals, span) {
  const k = 2 / (span + 1);
  const out = [vals[0]];
  for (let i = 1; i < vals.length; i++) {
    out.push(vals[i] * k + out[out.length - 1] * (1 - k));
  }
  return out;
}

function macd(arr) {
  const e12 = ema(arr, 12), e26 = ema(arr, 26);
  const line = e12.map((v, i) => v - e26[i]);
  const sig = ema(line, 9);
  return [line[line.length - 1], sig[sig.length - 1]];
}

function atr(highs, lows, closes, period = 14) {
  const trs = [];
  for (let i = 1; i < closes.length; i++) {
    const tr = Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1])
    );
    trs.push(tr);
  }
  if (trs.length < period) return null;
  return trs.slice(-period).reduce((a, b) => a + b, 0) / period;
}

const round = (x, d = 2) => {
  const m = Math.pow(10, d);
  return Math.round(x * m) / m;
};

// ───────────────────────── 評分 ─────────────────────────
async function screenOne(t, benchRet3m) {
  const { c, h, l, v } = await fetchYahoo(t);
  if (!c || c.length < 210) throw new Error("no data / <min bars");
  const ma50 = sma(c, 50), ma200 = sma(c, 200);
  const ma50s = smaSeries(c, 50), ma200s = smaSeries(c, 200);
  const r = rsi(c, 14);
  const [mline, msig] = macd(c);
  const cur = c[c.length - 1];
  const ret3m = c.length >= 63 ? (c[c.length - 1] / c[c.length - 63] - 1) * 100 : 0;
  const high52 = c.length >= 252 ? Math.max(...c.slice(-252)) : Math.max(...c);
  const distHigh = (cur / high52 - 1) * 100;
  const a = atr(h, l, c);
  const stop = a ? round(cur - 2.0 * a, 2) : null;
  const belowMa200 = ma200 ? cur < ma200 : false;

  let score = 0;
  if (ma50 && ma200 && ma50 > ma200) score += 2;
  if (r && r >= 40 && r <= 70) score += 1;
  if (mline > msig) score += 2;
  if (ret3m > benchRet3m) score += 1;
  if (distHigh >= -25.0 && distHigh < 0) score += 1;
  if (distHigh >= -15.0 && distHigh <= -3.0) score += 1;

  const W = 260;  // 畫圖窗口：夠長等 MA200 有連續線
  return {
    ticker: t.toUpperCase(),
    price: round(cur, 2),
    golden: !!(ma50 && ma200 && ma50 > ma200),
    ma50: ma50 ? round(ma50, 2) : null,
    ma200: ma200 ? round(ma200, 2) : null,
    rsi: r ? round(r, 1) : null,
    macd_bull: mline > msig,
    ret3m: round(ret3m, 1),
    rel_str: round(ret3m - benchRet3m, 1),
    dist_high: round(distHigh, 1),
    atr: a ? round(a, 2) : null,
    stop_loss: stop,
    below_ma200: belowMa200,
    score,
    series: c.slice(-W).map(x => round(x, 2)),
    volumes: v.slice(-W).map(x => Math.round(x)),
    ma50_series: ma50s.slice(-W),
    ma200_series: ma200s.slice(-W),
  };
}

async function runScreen(symbols) {
  let benchRet3m = 0.0;
  try {
    const bc = await fetchYahoo("SPY");
    benchRet3m = bc.c.length >= 63 ? (bc.c[bc.c.length - 1] / bc.c[bc.c.length - 63] - 1) * 100 : 0;
  } catch (_) { benchRet3m = 0.0; }

  const rows = [], errors = [];
  for (const s of symbols) {
    try {
      rows.push(await screenOne(s, benchRet3m));
    } catch (e) {
      errors.push({ ticker: s, error: String(e.message || e) });
    }
  }
  rows.sort((a, b) => b.score - a.score);
  return {
    rows,
    errors,
    bench_ret3m: round(benchRet3m, 1),
    count: rows.length,
  };
}

// ───────────────────────── Netlify Function entrypoint ─────────────────────────
export default async (req, context) => {
  try {
    let raw = "";
    if (req.method === "POST") {
      const body = await req.json().catch(() => ({}));
      raw = body.symbols || body.q || "";
    } else {
      const url = new URL(req.url);
      raw = url.searchParams.get("q") || url.searchParams.get("symbols") || "";
    }
    const symbols = raw.replace(/,/g, " ").split(/\s+/).map(s => s.trim().toUpperCase()).filter(Boolean).slice(0, 30);
    if (!symbols.length) {
      return new Response(JSON.stringify({ error: "請提供至少一個 symbol，例如 AAPL NVDA" }), {
        status: 400,
        headers: { "Content-Type": "application/json", "Access-Control-Allow-Origin": "*" },
      });
    }
    const data = await runScreen(symbols);
    return new Response(JSON.stringify(data), {
      status: 200,
      headers: { "Content-Type": "application/json", "Access-Control-Allow-Origin": "*" },
    });
  } catch (e) {
    return new Response(JSON.stringify({ error: `server error: ${e.message || e}` }), {
      status: 500,
      headers: { "Content-Type": "application/json", "Access-Control-Allow-Origin": "*" },
    });
  }
};

// ── 本地測試用 (node netlify/functions/screen.mjs "AAPL NVDA") ──
if (process.argv[1] && process.argv[1].endsWith("screen.mjs")) {
  const syms = process.argv[2] || "AAPL NVDA TSLA";
  runScreen(syms.split(/\s+/).map(s => s.trim().toUpperCase()))
    .then(d => { console.log(JSON.stringify(d, null, 2)); })
    .catch(e => { console.error("ERR", e); process.exit(1); });
}
